### ISSUE TITLE
Training on GPU with a custom container based on official pytorch-training container

### ISSUE THREAD

**Issue by jason-morgan** [aws/sagemaker-pytorch-container#166]
**TL;DR**

We train on a custom container based on the official `pytorch-training:1.3.1-gpu-py3` container. The training script runs just fine, but it doesn't utilize the GPU. The GPU Utilization according to CloudWatch sits at 0%, even though the GPU Memory is being used. My question: is there anything special we need to do to make sure the GPU is used?

And, yes, we use `.to('cuda:0')` liberally throughout the code, and verify with logging (see details).

Thanks!

**Details**

I am using a custom container to train a pytorch model. The custom `Dockerfile` uses the following base container:

`763104351884.dkr.ecr.us-east-2.amazonaws.com/pytorch-training:1.3.1-gpu-py3`

Here is the `Dockerfile` (cutting out some custom package installs):

```
FROM 763104351884.dkr.ecr.us-east-2.amazonaws.com/pytorch-training:1.3.1-gpu-py3
LABEL maintainer="Aware Behavioral Intelligence"
WORKDIR /

[snip]

COPY model /opt/ml/code
ENV PATH="/opt/ml/code:${PATH}"
ENV SAGEMAKER_SUBMIT_DIRECTORY /opt/ml/code
ENV SAGEMAKER_PROGRAM train
WORKDIR /opt/ml/code
```

As you can see from `SAGEMAKER_PROGRAM`, the script that is run is named `train`.

We submit our jobs using a `boto3.client(service_name='sagemaker')` client with the following (approximate) request:

```
response = client.create_training_job(
    TrainingJobName=SM_JOB_NAME,
    HyperParameters=HYPERPARAMETERS,
    AlgorithmSpecification={
        'TrainingImage': 'xxxxxxxxxxxx.dkr.ecr.us-east-2.amazonaws.com/sagemaker-cv-trainer',
        'TrainingInputMode': 'File',
        'MetricDefinitions': []
    },
    RoleArn=SM_ARN,
    InputDataConfig=[
        SM_TRAIN_INPUT,
        SM_DEV_INPUT
    ],
    OutputDataConfig=SM_OUTPUT,
    ResourceConfig={        
        'InstanceType': 'ml.p3.2xlarge',
        'InstanceCount': 1,
        'VolumeSizeInGB': 10
    },
    StoppingCondition={
        'MaxRuntimeInSeconds': 1*24*60*60
    },
    Tags=[
        {
            'Key': 'model',
            'Value': 'cv'
        }
    ]
)
```

Logging shows that pytorch detects the GPU:

![Screenshot from 2020-02-17 11-00-07](https://user-images.githubusercontent.com/4924143/74669305-e481f600-5174-11ea-9ac7-4820fe3586ae.png)

For each training batch / prediction batch, I make sure the data are on the GPU:

![Screenshot from 2020-02-17 11-02-33](https://user-images.githubusercontent.com/4924143/74669426-17c48500-5175-11ea-904d-ac334099a15a.png)

And data do appear to get moved to the GPU (but the GPU is otherwise not used):

![Screenshot from 2020-02-17 10-59-36](https://user-images.githubusercontent.com/4924143/74669501-3d518e80-5175-11ea-824f-4300d3729b4e.png)

Any insights into how we may be using Sagemaker incorrectly? We have been training and deploying models on Sagemaker using these same patterns for the last year, and they have worked fine, but now we are in need of GPU resources for computer vision tasks. 

Thank you.

**Comment by laurenyu** [comment]
since it looks like from the logs that `torch` is able to detect the GPU, I suspect there's something possibly missing in the training script.

are you sure that your model is using the GPU? if you're able to run your training script locally on a GPU instance, is the GPU utilized?

**Comment by jason-morgan** [comment]
Thank you, and I apologize for the delay in responding. We are not set up to train locally, so I had to mock some services before I could test out training on my laptop.

In the end, it appears that everything is working fine with respect to training on a GPU with the sagemaker pytorch containers. I was able to use both a custom container and the sagemaker pytorch container on my laptop to train on the GPU. It seems that 8 CPUs (for the GPU instance I am using) are just too few to keep the GPU busy (I am doing several steps of image augmentation during training). After fooling with batch sizes and number of Dataset workers, I was able to peak at an unsustained 25% of GPU utilization. I have had more success with my current code running on a larger CPU instance (ml.c5.9xlarge).
